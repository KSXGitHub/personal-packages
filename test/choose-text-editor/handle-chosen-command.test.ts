import {
  Command,
  CommandHandlingMethod,
  Status,
  SpawnSync,
  handleChosenCommand
} from '@khai96x/choose-text-editor'

import MockedConsole from './lib/mocked-console'

async function setup (
  handle: CommandHandlingMethod,
  spawnSyncImpl: SpawnSync = () => ({ status: 0 })
) {
  const args = ['a', 0, true, 'multiple words']
  const command: Command = {
    path: '/usr/bin/chosen-command',
    args: ['abc', 'def', 'multiple words again']
  }
  const console = new MockedConsole()
  const spawnSync = jest.fn(spawnSyncImpl)

  const result = await handleChosenCommand({
    handle,
    args,
    command,
    spawnSync,
    logInfo: console.info,
    logError: console.error
  })

  const status = Status[result]

  return {
    handle,
    args,
    command,
    console,
    spawnSync,
    result,
    status
  }
}

describe(`when handling method is ${CommandHandlingMethod.PrintSingleLine}`, () => {
  const handle = CommandHandlingMethod.PrintSingleLine

  it('calls logInfo', async () => {
    const { console } = await setup(handle)
    expect(console.getInfoLogs()).toMatchSnapshot()
  })

  it('does not call logError', async () => {
    const { console } = await setup(handle)
    expect(console.error).not.toBeCalled()
  })

  it('does not call spawnSync', async () => {
    const { spawnSync } = await setup(handle)
    expect(spawnSync).not.toBeCalled()
  })

  it('returns status of Success', async () => {
    const { status } = await setup(handle)
    expect(status).toBe(Status[Status.Success])
  })
})

describe(`when handling method is ${CommandHandlingMethod.PrintMultiLine}`, () => {
  const handle = CommandHandlingMethod.PrintMultiLine

  it('calls logInfo', async () => {
    const { console } = await setup(handle)
    expect(console.getInfoLogs()).toMatchSnapshot()
  })

  it('prints message matches snapshot', async () => {
    const { console } = await setup(handle)
    expect(console.getInfoText()).toMatchSnapshot()
  })

  it('does not call logError', async () => {
    const { console } = await setup(handle)
    expect(console.error).not.toBeCalled()
  })

  it('does not call spawnSync', async () => {
    const { spawnSync } = await setup(handle)
    expect(spawnSync).not.toBeCalled()
  })

  it('returns status of Success', async () => {
    const { status } = await setup(handle)
    expect(status).toBe(Status[Status.Success])
  })
})

describe(`when handling method is ${CommandHandlingMethod.PrintJson}`, () => {
  const handle = CommandHandlingMethod.PrintJson

  it('calls logInfo', async () => {
    const { console } = await setup(handle)
    expect(console.getInfoLogs()).toMatchSnapshot()
  })

  it('prints a valid JSON message', async () => {
    const { console } = await setup(handle)
    expect(JSON.parse(console.getInfoText())).toMatchSnapshot()
  })

  it('does not call logError', async () => {
    const { console } = await setup(handle)
    expect(console.error).not.toBeCalled()
  })

  it('does not call spawnSync', async () => {
    const { spawnSync } = await setup(handle)
    expect(spawnSync).not.toBeCalled()
  })

  it('returns status of Success', async () => {
    const { status } = await setup(handle)
    expect(status).toBe(Status[Status.Success])
  })
})

describe(`when handling method is ${CommandHandlingMethod.Execute}`, () => {
  const handle = CommandHandlingMethod.Execute

  describe('when spawnSync returns object with 0 as status code', () => {
    it('does not call logInfo', async () => {
      const { console } = await setup(handle)
      expect(console.info).not.toBeCalled()
    })

    it('does not call logError', async () => {
      const { console } = await setup(handle)
      expect(console.error).not.toBeCalled()
    })

    it('calls spawnSync', async () => {
      const { spawnSync } = await setup(handle)
      expect(spawnSync.mock.calls).toMatchSnapshot()
    })

    it('returns status of Success', async () => {
      const { status } = await setup(handle)
      expect(status).toBe(Status[Status.Success])
    })
  })

  describe('when spawnSync returns object with non-zero status', () => {
    const spawnSyncImpl: SpawnSync = () => ({ status: 123 })

    it('does not call logInfo', async () => {
      const { console } = await setup(handle, spawnSyncImpl)
      expect(console.info).not.toBeCalled()
    })

    it('calls logError', async () => {
      const { console } = await setup(handle, spawnSyncImpl)
      expect(console.getErrorLogs()).toMatchSnapshot()
    })

    it('prints error message matches snapshot', async () => {
      const { console } = await setup(handle, spawnSyncImpl)
      expect(console.getErrorText()).toMatchSnapshot()
    })

    it('calls spawnSync', async () => {
      const { spawnSync } = await setup(handle, spawnSyncImpl)
      expect(spawnSync.mock.calls).toMatchSnapshot()
    })

    it('returns status of ExecutionFailure', async () => {
      const { status } = await setup(handle, spawnSyncImpl)
      expect(status).toBe(Status[Status.ExecutionFailure])
    })
  })

  describe('when spawnSync returns object with non-null error', () => {
    class CustomError {
      errno = -2
      code = 'ENOENT'
      syscall = 'spawnSync'
      path = '{path}'
      spawnargs = []
    }

    const spawnSyncImpl: SpawnSync = () => ({
      status: null,
      error: new CustomError() as any
    })

    it('does not call logInfo', async () => {
      const { console } = await setup(handle, spawnSyncImpl)
      expect(console.info).not.toBeCalled()
    })

    it('calls logError', async () => {
      const { console } = await setup(handle, spawnSyncImpl)
      expect(console.getErrorLogs()).toMatchSnapshot()
    })

    it('prints error message matches snapshot', async () => {
      const { console } = await setup(handle, spawnSyncImpl)
      expect(console.getErrorText()).toMatchSnapshot()
    })

    it('calls spawnSync', async () => {
      const { spawnSync } = await setup(handle, spawnSyncImpl)
      expect(spawnSync.mock.calls).toMatchSnapshot()
    })

    it('returns status of ExecutionFailure', async () => {
      const { status } = await setup(handle, spawnSyncImpl)
      expect(status).toBe(Status[Status.ExecutionFailure])
    })
  })
})
