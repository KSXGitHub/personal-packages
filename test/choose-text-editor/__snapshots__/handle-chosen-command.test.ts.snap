// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when handling method is exec when spawnSync returns object with 0 as status code calls spawnSync 1`] = `
Array [
  Array [
    "/usr/bin/chosen-command",
    Array [
      "abc",
      "def",
      "multiple words again",
      "a",
      "0",
      "true",
      "multiple words",
    ],
    Object {
      "stdio": "inherit",
    },
  ],
]
`;

exports[`when handling method is exec when spawnSync returns object with non-null error calls logError 1`] = `
Array [
  Array [
    "[ERROR] Failed to execute command",
  ],
  Array [
    "* executable: '/usr/bin/chosen-command'",
  ],
  Array [
    "* error: CustomError {
  errno: -2,
  code: 'ENOENT',
  syscall: 'spawnSync',
  path: '{path}',
  spawnargs: []
}",
  ],
]
`;

exports[`when handling method is exec when spawnSync returns object with non-null error calls spawnSync 1`] = `
Array [
  Array [
    "/usr/bin/chosen-command",
    Array [
      "abc",
      "def",
      "multiple words again",
      "a",
      "0",
      "true",
      "multiple words",
    ],
    Object {
      "stdio": "inherit",
    },
  ],
]
`;

exports[`when handling method is exec when spawnSync returns object with non-null error prints error message matches snapshot 1`] = `
"
[ERROR] Failed to execute command
* executable: '/usr/bin/chosen-command'
* error: CustomError {
  errno: -2,
  code: 'ENOENT',
  syscall: 'spawnSync',
  path: '{path}',
  spawnargs: []
}
"
`;

exports[`when handling method is exec when spawnSync returns object with non-zero status calls logError 1`] = `
Array [
  Array [
    "[ERROR] Execution of command resulted in failure",
  ],
  Array [
    "* executable: '/usr/bin/chosen-command'",
  ],
  Array [
    "* arguments: [
  'abc',
  'def',
  'multiple words again',
  'a',
  0,
  true,
  'multiple words'
]",
  ],
  Array [
    "* status: 123",
  ],
]
`;

exports[`when handling method is exec when spawnSync returns object with non-zero status calls spawnSync 1`] = `
Array [
  Array [
    "/usr/bin/chosen-command",
    Array [
      "abc",
      "def",
      "multiple words again",
      "a",
      "0",
      "true",
      "multiple words",
    ],
    Object {
      "stdio": "inherit",
    },
  ],
]
`;

exports[`when handling method is exec when spawnSync returns object with non-zero status prints error message matches snapshot 1`] = `
"
[ERROR] Execution of command resulted in failure
* executable: '/usr/bin/chosen-command'
* arguments: [
  'abc',
  'def',
  'multiple words again',
  'a',
  0,
  true,
  'multiple words'
]
* status: 123
"
`;

exports[`when handling method is print:json calls logInfo 1`] = `
Array [
  Array [
    "{
  \\"path\\": \\"/usr/bin/chosen-command\\",
  \\"args\\": [
    \\"abc\\",
    \\"def\\",
    \\"multiple words again\\",
    \\"a\\",
    \\"0\\",
    \\"true\\",
    \\"multiple words\\"
  ]
}",
  ],
]
`;

exports[`when handling method is print:json prints a valid JSON message 1`] = `
Object {
  "args": Array [
    "abc",
    "def",
    "multiple words again",
    "a",
    "0",
    "true",
    "multiple words",
  ],
  "path": "/usr/bin/chosen-command",
}
`;

exports[`when handling method is print:multi calls logInfo 1`] = `
Array [
  Array [
    "/usr/bin/chosen-command",
  ],
  Array [
    "abc",
  ],
  Array [
    "def",
  ],
  Array [
    "'multiple words again'",
  ],
  Array [
    "a",
  ],
  Array [
    "0",
  ],
  Array [
    "true",
  ],
  Array [
    "'multiple words'",
  ],
]
`;

exports[`when handling method is print:multi prints message matches snapshot 1`] = `
"
/usr/bin/chosen-command
abc
def
'multiple words again'
a
0
true
'multiple words'
"
`;

exports[`when handling method is print:single calls logInfo 1`] = `
Array [
  Array [
    "/usr/bin/chosen-command abc def 'multiple words again' a 0 true 'multiple words'",
  ],
]
`;
